### PR TITLE
feat(core): add ensureStylableImports util

### PR DESCRIPTION
**Pull request by barak007**
This is WIP just so we can talk about where to put editing helper functions

### CODE REVIEW

File: packages/core/test/stylable-import-tools.spec.ts [hunk 1]
```diff
@@ -0,0 +1,466 @@
...
+            ensureStylableImports(root, [], { mode: ':import' }, diag);
+            ensureStylableImports(root, [], { mode: 'st-import' }, diag);
+
+            const importNode = root.nodes[0];
+            expect(diag.reports, 'No diagnostics').to.have.lengthOf(0);
+            expect(importNode.toString()).to.equal(`:import {-st-from: 'x'; -st-default: A }`);
+        });
+    });
+
+    describe('patch-only mode', () => {
```
Comment (idoros) [inline packages/core/test/stylable-import-tools.spec.ts:25]: missing case(s) that `patch-only` modifies `@st-import`

File: packages/core/src/stylable-imports-tools.ts [hunk 2]
```diff
@@ -0,0 +1,430 @@
...
+
+    params += ` from ${JSON.stringify(importObj.request || '')}`;
+    return { name: 'st-import', params };
+}
+
+export function ensureStylableImports(
+    ast: Root,
+    importPatches: Array<ImportPatch>,
+    options: {
+        mode: 'patch-only' | 'st-import' | ':import';
```
Comment (idoros) [inline packages/core/src/stylable-imports-tools.ts:91]: I'm not sure about the `mode` option. `patch-only` works on both types - that is the prefered mode, why give any other option? Maybe add a different option `newImportType` to handle addition of non existing imports (with default to `st-import`)

File: packages/core/src/stylable-imports-tools.ts [hunk 3]
```diff
@@ -0,0 +1,430 @@
...
+            parts.push(name);
+        } else {
+            parts.push(`${name} as ${as}`);
+        }
+    }
+
+    return parts;
+}
+
+type ImportPatch = {
```
Comment (idoros) [inline packages/core/src/stylable-imports-tools.ts:351]: I understand that the fields are copies of the already existing `Imported` interface, but `request` would have been more clear as `from`, `defaultExport` should have been `defaultImport`

File: packages/core/test/stylable-import-tools.spec.ts [hunk 4]
```diff
@@ -0,0 +1,466 @@
...
+            );
+
+            const importNode = root.nodes[0];
+            expect(diag.reports, 'No diagnostics').to.have.lengthOf(0);
+            expect(importNode.toString()).to.equal(`:import {-st-from: 'x';-st-default: Test; }`);
+        });
+
+        it('should add named import', () => {
+            const root = parse(`
+                :import {-st-from: 'x'; }
```
Comment (idoros) [inline packages/core/test/stylable-import-tools.spec.ts:63]: If you had 2 imports from the same `from` path, would it add the import symbols to both and cause a duplicate symbol error?

File: packages/core/src/stylable-imports-tools.ts [hunk 5]
```diff
@@ -0,0 +1,430 @@
...
+    ATTEMPT_OVERRIDE_SYMBOL(
+        kind: 'default' | 'named' | 'keyframes',
+        origin: string,
+        override: string
+    ) {
+        return `Attempt to override existing ${kind} import symbol. ${origin} -> ${override} `;
+    },
+};
+
+export function createAtImportProps(
```
[the remark this answers is not in the document]
Comment (idoros) [reply packages/core/src/stylable-imports-tools.ts:51]: I see now, CLI is importing through dist. We should talk about an internal index to pass exports between stylable packages.

File: packages/core/src/index.ts [hunk 6]
```diff
@@ -26,8 +26,10 @@ export {
     processNamespace,
     processorWarnings,
     validateScopingSelector,
-    parsePseudoImport,
 } from './stylable-processor';
+
+export { ensureStylableImports, parsePseudoImport, parseStImport } from './stylable-imports-tools';
```
[the remark this answers is not in the document]
Comment (barak007) [reply packages/core/src/index.ts:31]: I will still use the specific functions when needed

File: packages/core/test/stylable-import-tools.spec.ts [hunk 7]
```diff
@@ -0,0 +1,466 @@
...
+            expect(diagnostics.reports[0].message).to.equal(
+                ensureImportsMessages.ATTEMPT_OVERRIDE_SYMBOL('named', 'Y', 'X as Y')
+            );
+        });
+        it('should report collision diagnostics for named "as" with "as" (no patch)', () => {
+            const root = parse(`@st-import [A as Y] from "x";`);
+
+            const { diagnostics } = ensureStylableImports(
+                root,
+                [{ request: 'x', named: { Y: 'X' } }],
```
[the remark this answers is not in the document]
Comment (barak007) [reply packages/core/test/stylable-import-tools.spec.ts:434]: our named object is like {asName: symbolName}

File: packages/core/src/stylable-processor.ts [hunk 8]
```diff
@@ -69,12 +68,6 @@ export const processorWarnings = {
     ROOT_AFTER_SPACING() {
         return '".root" class cannot be used after native elements or selectors external to the stylesheet';
     },
-    DEFAULT_IMPORT_IS_LOWER_CASE() {
```
[the remark this answers is not in the document]
Comment (barak007) [reply packages/core/src/stylable-processor.ts:72]: if some one is...I have no problem to break him. this is a purely internal stuff.
we just need to remember to clean it up after/before merge

File: packages/core/src/stylable-imports-tools.ts (hunk 5 above)
[the remark this answers is not in the document]
Comment (barak007) [reply packages/core/src/stylable-imports-tools.ts:51]: yes we do. not in the PR

File: packages/core/src/stylable-imports-tools.ts [hunk 9]
```diff
@@ -0,0 +1,430 @@
...
+    importPatches: Array<ImportPatch>,
+    options: {
+        mode: 'patch-only' | 'st-import' | ':import';
+        shouldPatch?: (diagnostics: Diagnostics) => boolean;
+    },
+    diagnostics: Diagnostics = new Diagnostics()
+) {
+    const patches = createImportPatches(ast, importPatches, options, diagnostics);
+    const shouldPatch = options.shouldPatch || (() => !diagnostics.reports.length);
+    if (shouldPatch(diagnostics)) {
```
[the remark this answers is not in the document]
Comment (barak007) [reply packages/core/src/stylable-imports-tools.ts:98]: also I think I will remove it for now.

File: packages/core/src/stylable-imports-tools.ts [hunk 10]
```diff
@@ -0,0 +1,430 @@
...
+function hasDefinitions({
+    named = {},
+    keyframes = {},
+    defaultExport,
+}: Partial<Pick<Imported, 'named' | 'keyframes' | 'defaultExport'>>) {
+    return defaultExport || Object.keys(named).length || Object.keys(keyframes).length;
+}
+
+function processImports(
+    pseudoImport: Imported,
```
[the remark this answers is not in the document]
Comment (barak007) [reply packages/core/src/stylable-imports-tools.ts]: called it imported

File: packages/core/test/stylable-import-tools.spec.ts (hunk 4 above)
Comment (barak007) [reply packages/core/test/stylable-import-tools.spec.ts:63]: fixed

**Pull request merged**